File: 19/main.go
package main

import (
	"bufio"
	"fmt"
	"io"
	"os"
	"runtime"
	"time"
)

const MOD = 1000000007

// checkLimits проверяет ограничения времени и памяти (работает только если установлена переменная окружения CHECK_LIMITS)
func checkLimits(maxTime time.Duration, maxMemoryMB int, fn func()) {
	if os.Getenv("CHECK_LIMITS") == "" {
		fn()
		return
	}

	var m1, m2 runtime.MemStats
	runtime.GC()
	runtime.ReadMemStats(&m1)

	start := time.Now()
	fn()
	elapsed := time.Since(start)

	runtime.GC()
	runtime.ReadMemStats(&m2)

	allocated := m2.TotalAlloc - m1.TotalAlloc
	memoryMB := float64(allocated) / (1024 * 1024)
	maxMemoryBytes := uint64(maxMemoryMB) * 1024 * 1024

	timeOk := elapsed <= maxTime
	memoryOk := allocated <= maxMemoryBytes

	if !timeOk || !memoryOk {
		if elapsed > maxTime {
			fmt.Fprintf(os.Stderr, "⚠️ Превышено время: %v (лимит: %v)\n", elapsed, maxTime)
		}
		if memoryMB > float64(maxMemoryMB) {
			fmt.Fprintf(os.Stderr, "⚠️ Превышена память: %.2f МБ (лимит: %d МБ)\n", memoryMB, maxMemoryMB)
		}
	} else {
		fmt.Fprintf(os.Stderr, "✓ Время: %v, Память: %.2f МБ\n", elapsed, memoryMB)
	}
}

type FastScanner struct {
	r    io.Reader
	buf  []byte
	pos  int
	size int
}

func NewFastScanner(r io.Reader) *FastScanner {
	return &FastScanner{
		r:   r,
		buf: make([]byte, 1<<16), // 64KB buffer
	}
}

func (s *FastScanner) NextInt() int {
	var n int
	// Skip whitespace
	for {
		if s.pos >= s.size {
			var err error
			s.size, err = s.r.Read(s.buf)
			s.pos = 0
			if err != nil || s.size == 0 {
				return 0
			}
		}
		c := s.buf[s.pos]
		s.pos++
		if c >= '0' && c <= '9' {
			n = int(c - '0')
			break
		}
	}
	// Read number
	for {
		if s.pos >= s.size {
			var err error
			s.size, err = s.r.Read(s.buf)
			s.pos = 0
			if err != nil || s.size == 0 {
				break
			}
		}
		c := s.buf[s.pos]
		if c < '0' || c > '9' {
			break
		}
		s.pos++
		n = n*10 + int(c-'0')
	}
	return n
}

func main() {
	checkLimits(1*time.Second, 256, solve)
}

func solve() {
	reader := bufio.NewReaderSize(os.Stdin, 1<<20)
	writer := bufio.NewWriterSize(os.Stdout, 1<<20)
	defer writer.Flush()

	scanner := NewFastScanner(reader)

	m := scanner.NextInt()
	if m == 0 {
		return
	}

	var N int64 = 0
	var lastCount int64 = 0

	for i := 0; i < m; i++ {
		val := int64(scanner.NextInt())
		N += val
		if val > 0 {
			lastCount = val
		}
	}

	if N == 0 {
		fmt.Fprintln(writer, 0)
		return
	}

	S := N - lastCount
	if S == 0 {
		fmt.Fprintln(writer, 0)
		return
	}

	// Предвычисление обратных элементов
	inv := make([]int64, N+1)
	inv[1] = 1
	for i := int64(2); i <= N; i++ {
		inv[i] = (MOD - (MOD/i)*inv[MOD%i]%MOD) % MOD
	}

	var sumInv int64 = 0

	// Сумма 1/(k(N-k)) для k от 1 до S
	for k := int64(1); k <= S; k++ {
		// term = 1/(k*(N-k)) = inv[k] * inv[N-k]
		term := (inv[k] * inv[N-k]) % MOD
		sumInv = (sumInv + term) % MOD
	}

	// Coeff = N*(N-1)/2
	coeff := (N * (N - 1)) % MOD
	coeff = (coeff * inv[2]) % MOD

	ans := (coeff * sumInv) % MOD
	fmt.Fprintln(writer, ans)
}

File: 19/main.rs
use std::env;
use std::io::{self, Read, BufWriter, Write};
use std::time::Instant;

const MOD: i64 = 1_000_000_007;

// get_memory_usage возвращает текущее использование памяти в байтах (приблизительно)
fn get_memory_usage() -> Option<u64> {
    #[cfg(target_os = "linux")]
    {
        use std::fs;
        if let Ok(content) = fs::read_to_string("/proc/self/status") {
            for line in content.lines() {
                if line.starts_with("VmRSS:") {
                    if let Some(value) = line.split_whitespace().nth(1) {
                        if let Ok(kb) = value.parse::<u64>() {
                            return Some(kb * 1024);
                        }
                    }
                }
            }
        }
    }
    
    #[cfg(target_os = "macos")]
    {
        use std::process::Command;
        let pid = std::process::id().to_string();
        if let Ok(output) = Command::new("ps")
            .args(&["-o", "rss=", "-p", &pid])
            .output()
        {
            if let Ok(mem_str) = String::from_utf8(output.stdout) {
                if let Ok(kb) = mem_str.trim().parse::<u64>() {
                    return Some(kb * 1024);
                }
            }
        }
    }
    
    None
}

// check_limits проверяет ограничения времени и памяти
fn check_limits(max_time_ms: u64, max_memory_mb: u64, f: impl FnOnce()) {
    if env::var("CHECK_LIMITS").is_err() {
        f();
        return;
    }

    let mem_before = get_memory_usage();
    let start = Instant::now();
    f();
    let elapsed = start.elapsed();
    let mem_after = get_memory_usage();

    let elapsed_ms = elapsed.as_millis() as u64;
    let time_ok = elapsed_ms <= max_time_ms;

    let memory_ok = if let (Some(before), Some(after)) = (mem_before, mem_after) {
        let allocated = if after > before { after - before } else { 0 };
        let max_memory_bytes = max_memory_mb * 1024 * 1024;
        allocated <= max_memory_bytes
    } else {
        true
    };

    if !time_ok || !memory_ok {
        if elapsed_ms > max_time_ms {
            eprintln!("⚠️ Превышено время: {} мс (лимит: {} мс)", elapsed_ms, max_time_ms);
        }
        if let (Some(before), Some(after)) = (mem_before, mem_after) {
            let allocated = if after > before { after - before } else { 0 };
            let memory_mb = allocated as f64 / (1024.0 * 1024.0);
            if memory_mb > max_memory_mb as f64 {
                eprintln!("⚠️ Превышена память: {:.2} МБ (лимит: {} МБ)", memory_mb, max_memory_mb);
            }
        }
    } else {
        if let (Some(before), Some(after)) = (mem_before, mem_after) {
            let allocated = if after > before { after - before } else { 0 };
            let memory_mb = allocated as f64 / (1024.0 * 1024.0);
            eprintln!("✓ Время: {} мс, Память: {:.2} МБ", elapsed_ms, memory_mb);
        } else {
            eprintln!("✓ Время: {} мс, Память: не удалось измерить", elapsed_ms);
        }
    }
}

struct Scanner<R> {
    reader: R,
    buffer: Vec<u8>,
    pos: usize,
    cap: usize,
}

impl<R: Read> Scanner<R> {
    fn new(reader: R) -> Self {
        Self {
            reader,
            buffer: vec![0; 1 << 16],
            pos: 0,
            cap: 0,
        }
    }

    fn next_i64(&mut self) -> i64 {
        let mut n = 0i64;
        loop {
            if self.pos >= self.cap {
                self.cap = self.reader.read(&mut self.buffer).unwrap_or(0);
                self.pos = 0;
                if self.cap == 0 {
                    return 0;
                }
            }
            let c = self.buffer[self.pos];
            self.pos += 1;
            if c >= b'0' && c <= b'9' {
                n = (c - b'0') as i64;
                break;
            }
        }
        loop {
            if self.pos >= self.cap {
                self.cap = self.reader.read(&mut self.buffer).unwrap_or(0);
                self.pos = 0;
                if self.cap == 0 {
                    break;
                }
            }
            let c = self.buffer[self.pos];
            if c < b'0' || c > b'9' {
                break;
            }
            self.pos += 1;
            n = n * 10 + (c - b'0') as i64;
        }
        n
    }
}

fn solve() {
    let stdin = io::stdin();
    let stdout = io::stdout();
    let mut writer = BufWriter::new(stdout.lock());
    let mut scanner = Scanner::new(stdin.lock());

    let m = scanner.next_i64();
    if m == 0 {
        return;
    }

    let mut n: i64 = 0;
    let mut last_count: i64 = 0;

    for _ in 0..m {
        let val = scanner.next_i64();
        n += val;
        if val > 0 {
            last_count = val;
        }
    }

    if n == 0 {
        writeln!(writer, "0").unwrap();
        return;
    }

    let s = n - last_count;
    if s == 0 {
        writeln!(writer, "0").unwrap();
        return;
    }

    // Предвычисление обратных элементов
    let mut inv = vec![0i64; (n + 1) as usize];
    inv[1] = 1;
    for i in 2..=n {
        inv[i as usize] = (MOD - (MOD / i) * inv[(MOD % i) as usize] % MOD) % MOD;
    }

    let mut sum_inv: i64 = 0;

    // Сумма 1/(k(N-k)) для k от 1 до S
    for k in 1..=s {
        // term = 1/(k*(N-k)) = inv[k] * inv[N-k]
        let term = (inv[k as usize] * inv[(n - k) as usize]) % MOD;
        sum_inv = (sum_inv + term) % MOD;
    }

    // Coeff = N*(N-1)/2
    let mut coeff = (n * (n - 1)) % MOD;
    coeff = (coeff * inv[2]) % MOD;

    let ans = (coeff * sum_inv) % MOD;
    writeln!(writer, "{}", ans).unwrap();
    writer.flush().unwrap();
}

fn main() {
    check_limits(1000, 256, solve);
}

